File: iosApp/iosApp/Modules/CurrencyPicker/CurrencyPickerWireframeFactory.swift
// Created by web3d4v on 06/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - CurrencyPickerWireframeFactory

protocol CurrencyPickerWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: CurrencyPickerWireframeContext
    ) -> CurrencyPickerWireframe
}

// MARK: - DefaultCurrencyPickerWireframeFactory

final class DefaultCurrencyPickerWireframeFactory {
    private let currencyAddWireframeFactory: CurrencyAddWireframeFactory
    private let walletService: WalletService
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    
    init(
        currencyAddWireframeFactory: CurrencyAddWireframeFactory,
        walletService: WalletService,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.currencyAddWireframeFactory = currencyAddWireframeFactory
        self.walletService = walletService
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultCurrencyPickerWireframeFactory: CurrencyPickerWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: CurrencyPickerWireframeContext
    ) -> CurrencyPickerWireframe {
        DefaultCurrencyPickerWireframe(
            parent,
            context: context,
            currencyAddWireframeFactory: currencyAddWireframeFactory,
            walletService: walletService,
            networksService: networksService,
            currencyStoreService: currencyStoreService
        )
    }
}

// MARK: - Assembler

final class CurrencyPickerWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> CurrencyPickerWireframeFactory in
            DefaultCurrencyPickerWireframeFactory(
                currencyAddWireframeFactory: resolver.resolve(),
                walletService: resolver.resolve(),
                networksService: resolver.resolve(),
                currencyStoreService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/currencypicker/CurrencyPickerWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.currencypicker

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframe
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframeContext
import smartadapter.internal.extension.name

interface CurrencyPickerWireframeFactory {
    fun make(parent: Fragment?, context: CurrencyPickerWireframeContext): CurrencyPickerWireframe
}

class DefaultCurrencyPickerWireframeFactory(
    private val walletService: WalletService,
    private val currencyStoreService: CurrencyStoreService,
): CurrencyPickerWireframeFactory {

    override fun make(
        parent: Fragment?, context: CurrencyPickerWireframeContext
    ): CurrencyPickerWireframe = DefaultCurrencyPickerWireframe(
        parent?.let { WeakRef(it) },
        context,
        walletService,
        currencyStoreService,
    )
}

class CurrencyPickerWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(CurrencyPickerWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultCurrencyPickerWireframeFactory(
                it.resolve(WalletService::class.name),
                it.resolve(CurrencyStoreService::class.name),
            )
        }
    }
}